### PR TITLE
Encrypt and decrypt with Cryptography

### PR DESCRIPTION
**Pull request by jpmelos**
Related to https://github.com/closeio/flask-common/issues/63.
Closes https://github.com/closeio/flask-common/issues/65.

This depends on merging https://github.com/closeio/flask-common/pull/70 first, then merging another intermediary PR that adds version 0 to current data. Then, we can merge this. After merging this, we have another PR to drop treatment for version 0.

Decrypts version `0` and `1` data
Encrypts version `1` data

### CODE REVIEW

**Comment by tsx** [comment]
Can we make it work on py3 (the crypto functions themselves at least)? Otherwise we'll have to revisit it again pretty soon.

**Comment by jpmelos** [comment]
> Can we make it work on py3 (the crypto functions themselves at least)? Otherwise we'll have to revisit it again pretty soon.

This is something I talked about with @jkemp101 and we decided to keep this off the scope of this ticket, at the time this was proving to be a little bit hard. I can try again, `cryptography` may have fixed a few issues that `pycrypto` and `pycryptodome` had. I'll take a quick look, but if I think this could take a long time, I'll leave this off the scope.

File: flask_common/crypto.py [hunk 1]
```diff
@@ -32,41 +43,52 @@ def aes_generate_key():
...
+    return V1_MARKER + iv + aes_encrypt_iv(key, data, iv)
 
 
 # Verify + decrypt data encrypted with IV
 def aes_decrypt(key, data):
     assert len(key) == KEY_LENGTH, 'invalid key size'
-    iv = data[:AES_BLOCK_SIZE]
-    data = data[AES_BLOCK_SIZE:]
-    return aes_decrypt_iv(key, data, iv)
+    is_v0 = False
```
Comment (jkemp101) [inline flask_common/crypto.py]: I don't think we should have `is_v?` variables. We should just extract the version and pass it to all functions so they can act on it as they see fit.

File: flask_common/crypto.py [hunk 2]
```diff
@@ -38,51 +43,50 @@ def aes_generate_key():
...
 # Encrypt + sign using no IV or provided IV. Pass empty string for no IV.
 # Note: You should normally use aes_encrypt()
 def aes_encrypt_iv(key, data, iv):
-    aes_key = key[:AES_BLOCK_SIZE]
-    hmac_key = key[AES_BLOCK_SIZE:]
-    initial_value = long(iv.encode("hex"), 16) if iv else 1
-    ctr = Counter.new(128, initial_value=initial_value)
-    cipher = AES.new(aes_key, AES.MODE_CTR, counter=ctr).encrypt(data)
+    aes_key = key[:AES_KEY_SIZE]
+    hmac_key = key[AES_KEY_SIZE:]
```
[the remark this answers is not in the document]
Comment (jpmelos) [reply flask_common/crypto.py]: No, this is correct.

When we generate the key, we do:

```python
AES_KEY_SIZE = 32  # 256 bits
HMAC_KEY_SIZE = 32  # 256 bits
KEY_LENGTH = AES_KEY_SIZE + HMAC_KEY_SIZE

def aes_generate_key():
    return os.urandom(KEY_LENGTH)
```

Which means the `key` is the length of the AES key + HMAC key. So we are basically saying:

```python
aes_key = key[:AES_KEY_SIZE]
hmac_key = key[AES_KEY_SIZE:] 
```

The AES key is `key` up to `AES_KEY_SIZE`. The HMAC key is "whatever comes after". The equivalent would be `hmac_key = key[-HMAC_KEY_SIZE:]`, but doesn't really make a difference in our case. Both are conceptually correct and would work even if the keys had different sizes.

File: flask_common/crypto.py [hunk 3]
```diff
@@ -38,51 +43,50 @@ def aes_generate_key():
...
 
 
 # Verify + decrypt using no IV or provided IV. Pass empty string for no IV.
 # Note: You should normally use aes_decrypt()
-def aes_decrypt_iv(key, data, iv, extracted_version=None):
-    aes_key = key[:AES_BLOCK_SIZE]
-    hmac_key = key[AES_BLOCK_SIZE:]
+def aes_decrypt_iv(key, data, iv, extracted_version):
+    aes_key = key[:AES_KEY_SIZE]
+    hmac_key = key[AES_KEY_SIZE:]
```
[the remark this answers is not in the document]
Comment (jpmelos) [reply flask_common/crypto.py]: See comment above.

File: flask_common/crypto.py [hunk 4]
```diff
@@ -38,51 +43,50 @@ def aes_generate_key():
...
+    if extracted_version == V0_MARKER:
+        iv = data[:AES_KEY_SIZE]
+        data = data[AES_KEY_SIZE:]
+    else:
+        iv = data[:IV_SIZE]
+        data = data[IV_SIZE:]
     return aes_decrypt_iv(key, data, iv, extracted_version)
 
 
 # Encrypt + sign using no IV or provided IV. Pass empty string for no IV.
```
[the remark this answers is not in the document]
Comment (jpmelos) [reply flask_common/crypto.py]: Yes, you are right. Fixed the comment instead of allowing no IV, since that is not safe.

**Comment by jpmelos** [comment]
@tsx our encrypt/decrypt functions work in Python 3 now. 🎉 cc @jkemp101

**Comment by jpmelos** [comment]
@jkemp101 once https://github.com/closeio/flask-common/pull/75 is approved, I'll merge that into `master` and then `master` into this one, resolve conflicts, and fix tests. Then we should be good to go.

**Comment by jpmelos** [comment]
@tsx sorry for requesting yet another review for this PR! I hope this one will be the last one. But, as a bonus, we will get to close https://github.com/closeio/flask-common/issues/65. 😄

**Comment by jpmelos** [comment]
@tsx (Only last commit has changes you haven't reviewed yet.)

File: flask_common/mongo/fields/crypto.py [hunk 5]
```diff
@@ -51,3 +52,18 @@ def to_python(self, value):
...
+class EncryptedStringField(EncryptedBinaryField):
+    """
+    Encrypted Unicode string field. Encryption is completely transparent
+    to the caller as the field automatically decrypts when the field
+    is accessed and encrypts when the document is saved. The
+    underlying algorithm currently is AES-256.
+    """
+
+    def to_python(self, value):
+        return value and self._decrypt(value).decode('utf-8') or None
```
Comment (tsx) [inline flask_common/mongo/fields/crypto.py]: I know it's been that way here but I prefer an actual ternary `process(value) if value else None` over `value and process(value) or None` as ternary is much more readable and clear in what the intention is.

File: flask_common/mongo/fields/crypto.py [hunk 6]
```diff
@@ -48,10 +48,14 @@ def _decrypt(self, data):
         raise AuthenticationError('message authentication failed')
 
     def to_python(self, value):
-        return value and self._decrypt(value) or None
+        if not value:
+            return None
+        return self._decrypt(value)
```
Comment (jkemp101) [inline flask_common/mongo/fields/crypto.py]: I think @tsx was suggesting something like
`return self._decrypt(value) if value else None`

Comment (jpmelos) [reply flask_common/mongo/fields/crypto.py]: Oops, you're right! Done in https://github.com/closeio/flask-common/pull/71/commits/96a4def567e3725bd0858f1b2ce02c8501db9791.

**Pull request merged**